File: 30days-of-code/20_sort.cc
#include <vector>
#include <iostream>
#include <algorithm>
#include <functional>

using namespace std;

using SortFunction = function<pair<vector<int>,size_t>(const vector<int>)>;


template<typename T>
pair<vector<T>,size_t> bubbleSortOutOfPlace(vector<T> list) {
    size_t swapCounter{};
    size_t totalSwaps{0};

    for (size_t i{0}; i<list.size(); ++i) {
        swapCounter = 0;
        for(auto it = std::begin(list)+1; it != std::end(list); ++it) {
            if (*(it-1) > *it) {
               iter_swap(it-1, it);
               swapCounter += 1;
            }
        }
        totalSwaps += swapCounter;
        if (swapCounter == 0) {
            break;
        }
    }
    return make_pair(list, totalSwaps);
}

template<typename T>
vector<T> readMany() {
    size_t N{};
    cin >> N;

    vector<T> in_list{};
    for (size_t i{0}; i<N; ++i) {
        T element{};
        cin >> element;
        
        in_list.push_back(element);
    }

    return in_list;
}

int main() {
    // readMany | sort | show
    vector<int> numbers{ readMany<int>() };

    SortFunction mySort = bubbleSortOutOfPlace<int>;
    pair<vector<int>, size_t> result{};

    result = mySort(numbers);
    vector<int> sortedNumbers = result.first;
    size_t totalSwaps = result.second;

    cout <<"Array is sorted in " <<totalSwaps <<" swaps." <<endl;
    cout <<"First Element: " <<sortedNumbers.front() <<endl;
    cout <<"Last Element: " <<sortedNumbers.back() <<endl;
    return 0;
}

File: 30days-of-code/20_sort.py
#!/bin/python3


def bubble_sort_outofplace(in_numbers):
    # no in-place sort
    # copy the list first
    numbers = in_numbers[:]
    list_size = len(numbers)
    total_swaps = 0

    for i in range(list_size):
        swaps_counter = 0
        for j in range(1, list_size):
            if numbers[j-1] > numbers[j]:
                numbers[j], numbers[j-1] = numbers[j-1], numbers[j]
                swaps_counter += 1
        total_swaps += swaps_counter
        if swaps_counter == 0:
            break

    return numbers, total_swaps


def read_input():
    n = int(input().strip())
    return list(map(int, input().strip().split(' ')))


if __name__ == "__main__":
    # this is the global namespace !!
    in_numbers = read_input()

    bubble_sort = bubble_sort_outofplace
    sorted_numbers, num_swaps = bubble_sort(in_numbers)

    print(f"Array is sorted in {num_swaps} swaps.")
    print(f"First Element: {sorted_numbers[0]}")
    print(f"Last Element: {sorted_numbers[-1]}")
